File: javascript/hangout.js
const callInPizzaDelivery = (pizza) => {
  console.log(`calling in pizza ${pizza}`);
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('pizza delivered');
      resolve(pizza);
    }, 5000);
  });
}

const inviteFriendOver = (friend) => {
  console.log(`calling friend ${friend}`);
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('friend arrived');
      resolve(friend);
    }, 3000);
  });
}

const hangOut = () => {
  const pizzaOrderPromise = callInPizzaDelivery('pepperoni');
  const friendInvitePromise = inviteFriendOver('Jack');
  Promise.all([pizzaOrderPromise, friendInvitePromise]).then((results) => {
    const [pizza, friend] = results;
    console.log(`Hanging out with ${friend} eating a ${pizza} pizza.`);
  });
  console.log('waiting for food and friend');
}

hangOut();

File: scala/hangout/src/main/scala/HangOut.scala
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent._

object HangOut extends App {

   def callInPizzaDelivery(pizza: String): Future[String] = {
    println(s"calling in pizza $pizza")
    Future {
      blocking {
        Thread.sleep(5000)
        pizza
      }
    }
  }

  def inviteFriendOver(friend: String): Future[String] = {
    println(s"calling friend $friend")
    Future {
      blocking {
        Thread.sleep(5000)
        friend
      }
    }
  }

  def hangout(): Unit = {
    val pizzaFuture = callInPizzaDelivery("pepperoni")
    val friendFuture = inviteFriendOver("Jack")
    for {
      pizza <- pizzaFuture
      friend <- friendFuture
    } println(s"Hanging out with $friend eating a $pizza pizza.")
  }

  hangout()
  scala.io.StdIn.readChar()
}
